### PR TITLE
v1 and v2 specs

### PR DESCRIPTION
**Pull request by d-v-b**
I copy+pasted v1 and v2 from zarr-python. Is there any reason not to keep all the specs in one place?

Also, I added some QOL stuff to gitignore.

### CODE REVIEW

**Comment by MSanKeys963** [comment]
Hi @d-v-b.

Thanks for working on this. It'd be good to keep specs in a single place.
I think we should also remove the v1 and v2 specs from https://zarr.readthedocs.io/ and hyperlink them to the new page on https://zarr-specs.readthedocs.io/.

**Comment by joshmoore** [comment]
What's the next step here, @MSanKeys963 & @d-v-b? Get this in and then remove from zarr-python? Both at the same time?

**Comment by d-v-b** [comment]
I think we could add them here first, then remove them from zarr python + update the relevant links?

**Comment by jbms** [comment]
Since there are many links to the old v2 location (I don't know about v1) it would be nice to keep them working somehow.

**Comment by joshmoore** [comment]
:100: Big fan of not breaking URLs.

File: docs/v1/v1.0.rst [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        "zarr_format": 1
+    }
+
+Data type encoding
+~~~~~~~~~~~~~~~~~~
+
+Simple data types are encoded within the array metadata resource as a
+string, following the `NumPy array protocol type string (typestr)
+format
+<numpy:arrays.interface>`_. The
```
Comment (MSanKeys963) [inline docs/v1/v1.0.rst:111]:
```suggestion
string, following the `NumPy array protocol type string (typestr)
format
<https://numpy.org/doc/stable/reference/arrays.interface.html>`_. The
```

This was giving 404. I've updated the link.

File: docs/v1/v1.0.rst [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+the data (``<``: little-endian, ``>``: big-endian, ``|``:
+not-relevant), a character code giving the basic type of the array,
+and an integer providing the number of bytes the type uses. The byte
+order MUST be specified. E.g., ``"<f8"``, ``">i4"``, ``"|b1"`` and
+``"|S12"`` are valid data types.
+
+Structure data types (i.e., with multiple named fields) are encoded as
+a list of two-element lists, following `NumPy array protocol type
+descriptions (descr)
+<numpy:arrays.interface>`_.
```
Comment (MSanKeys963) [inline docs/v1/v1.0.rst:122]:
```suggestion
a list of two-element lists, following `NumPy array protocol type
descriptions (descr)
<https://numpy.org/doc/stable/reference/arrays.interface.html>`_.
```

This was also giving 404. Updated the link.

File: docs/v1/v1.0.rst [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+Chunks
+------
+
+Each chunk of the array is compressed by passing the raw bytes for the
+chunk through the primary compression library to obtain a new sequence
+of bytes comprising the compressed chunk data. No header is added to
+the compressed bytes or any other modification made. The internal
+structure of the compressed bytes will depend on which primary
+compressor was used. For example, the `Blosc compressor
+<https://github.com/Blosc/c-blosc/blob/main/README_HEADER.rst>`_
```
Comment (MSanKeys963) [inline docs/v1/v1.0.rst:136]:
```suggestion
compressor was used. For example, the `Blosc compressor
<https://github.com/Blosc/c-blosc/blob/main/README_CHUNK_FORMAT.rst>`_
```
The `README_HEADER.rst` has been renamed to `README_CHUNK_FORMAT.rst`.

File: docs/v2/v2.0.rst [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+    }
+
+.. _spec_v2_array_dtype:
+
+Data type encoding
+~~~~~~~~~~~~~~~~~~
+
+Simple data types are encoded within the array metadata as a string,
+following the :ref:`NumPy array protocol type string (typestr) format
+<numpy:arrays.interface>`. The format
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:132]:
```suggestion
following the `NumPy array protocol type string (typestr) format <https://numpy.org/doc/stable/reference/arrays.interface.html>`_. The format
```
Fixed broken link.

File: docs/v2/v2.0.rst [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+  fixed-size chunk of memory))
+* An integer specifying the number of bytes the type uses.
+
+The byte order MUST be specified. E.g., ``"<f8"``, ``">i4"``, ``"|b1"`` and
+``"|S12"`` are valid data type encodings.
+
+For datetime64 ("M") and timedelta64 ("m") data types, these MUST also include the
+units within square brackets. A list of valid units and their definitions are given in
+the :ref:`NumPy documentation on Datetimes and Timedeltas
+<numpy:arrays.dtypes.dateunits>`.
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:151]:
```suggestion
the `NumPy documentation on Datetimes and Timedeltas <https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units>`_.
```

File: docs/v2/v2.0.rst [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+For example, ``"<M8[ns]"`` specifies a datetime64 data type with nanosecond time units.
+
+Structured data types (i.e., with multiple named fields) are encoded
+as a list of lists, following :ref:`NumPy array protocol type descriptions
+(descr)
+<numpy:arrays.interface>`. Each
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:157]:
```suggestion
as a list of lists, following `NumPy array protocol type descriptions
(descr) <https://numpy.org/doc/stable/reference/arrays.interface.html#the-array-interface-protocol>`_. Each
```

File: docs/v2/v2.0.rst [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+
+Version 2 clarifications
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following changes have been made to the version 2 specification since it was
+initially published to clarify ambiguities and add some missing information.
+
+* The specification now describes how bytes fill values should be encoded and
+  decoded for arrays with a fixed-length byte string data type (`#165
+  <https://github.com/zarr-developers/zarr-python/issue/165/>`_, `#176
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:535]:
```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/165/>`_, `#176
```

File: docs/v2/v2.0.rst [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+  <https://github.com/zarr-developers/zarr-python/issue/176/>`_).
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:536]:
```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/176/>`_).
```

File: docs/v2/v2.0.rst [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+
+* The specification now clarifies that units must be specified for datetime64 and
+  timedelta64 data types (`#85
+  <https://github.com/zarr-developers/zarr-python/issue/85/>`_, `#215
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:540]:
```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/85/>`_, `#215
```

File: docs/v2/v2.0.rst [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+  <https://github.com/zarr-developers/zarr-python/issue/215/>`_).
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:541]:
```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/215/>`_).
```

File: docs/v2/v2.0.rst [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+
+* The specification now clarifies that the '.zattrs' key does not have to be present for
+  either arrays or groups, and if absent then custom attributes should be treated as
+  empty.
+
+* The specification now describes how structured datatypes with
+  subarray shapes and/or with nested structured data types are encoded
+  in array metadata (`#111
+  <https://github.com/zarr-developers/zarr-python/issue/111/>`_, `#296
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:550]:
```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/111/>`_, `#296
```

File: docs/v2/v2.0.rst [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+  <https://github.com/zarr-developers/zarr-python/issue/296/>`_).
```
Comment (MSanKeys963) [inline docs/v2/v2.0.rst:551]:
```suggestion
  <https://github.com/zarr-developers/zarr-python/issues/296/>`_).
```

**Comment by MSanKeys963** [comment]
I've reviewed this PR thoroughly for any broken links and added https://github.com/zarr-developers/zarr-python/pull/1582 to remove spec v1 and v2.

I'd agree with @d-v-b to merge this PR first and then review and merge the https://github.com/zarr-developers/zarr-python/pull/1582.

**Comment by MSanKeys963** [comment]
@d-v-b - if you could please look at my review, we can go ahead with this PR soon.

After this is merged, we can also go ahead with https://github.com/zarr-developers/zarr-python/pull/1582 as well.

**Comment by MSanKeys963** [comment]
Both the [v1](https://zarr-specs--276.org.readthedocs.build/en/276/v1/v1.0.html) and [v2](https://zarr-specs--276.org.readthedocs.build/en/276/v2/v2.0.html) specs looks good now.

Thanks @jbms!

**Review by jhamman — APPROVED** [review]

**Pull request merged**